### PR TITLE
input type number in edition form when datatype is number

### PR DESCRIPTION
**Pull request by nworr**
Overload the input_htmlbootstrapFormWidget : outputContol method
Check if datatype is integer and use input type : number
by default, input type is text

browser will perform live validation and add spin button
Ticket : #2992 
 
input_number

### CODE REVIEW

**Comment by Gustry** [comment]
Nice addition !
Which option does it support from QGIS UI about this widget ? (min, max, step, suffix ?) I can see "max" from your PR.
Would it close #2992 ?

**Comment by Gustry** [comment]
A few GH Action did not run, maybe because of the branch ?

**Review by laurentj — APPROVED** [review]
lgtm :-)

**Comment by nworr** [comment]
@Gustry > currently when these options are set in QGIS UI , min & max  values are used for form validation only, 
suffix and step are ignored

but min/max/step can be used to setup html input form ,I'm working on it (on number only) , see (https://caniuse.com/input-number)

> Currently no mobile browsers and very few desktop browsers support using commas for languages where commas are used as decimal separators.

File: lizmap/plugins/formwidget/input_htmlbootstrap/input_htmlbootstrap.formwidget.php [hunk 1]
```diff
@@ -8,9 +8,32 @@
...
+        }
+        $maxl = $this->ctrl->datatype->getFacet('maxLength');
+        if ($maxl !== null) {
+            $attr['maxlength'] = $maxl;
+        }
+        $attr['value'] = $this->getValue();
+        // new code : check datatype to provide a better input type
+        if ( $this->ctrl->datatype instanceof jDatatypeInteger) {
+            $attr['type'] = 'number';
+        } else {
```
Comment (rldhont) [inline lizmap/plugins/formwidget/input_htmlbootstrap/input_htmlbootstrap.formwidget.php:31]:
```suggestion
        if ( $this->ctrl->datatype instanceof jDatatypeDecimal) {
            // Input Number
            // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/number
            $attr['type'] = 'number';
            $minValue = $this->ctrl->datatype->getFacet('minValue');
            if ($minValue !== null) {
                $attr['min'] = $minValue;
            }
            $maxValue = $this->ctrl->datatype->getFacet('maxValue');
            if ($maxValue !== null) {
                $attr['max'] = $maxValue;
            }
            // Allow decimal value
            // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/number#allowing_decimal_values
            if ( !($this->ctrl->datatype instanceof jDatatypeInteger) ) {
                $attr['step'] = 0.1;
            }
        } else {
```

**Review by rldhont — CHANGES_REQUESTED** [review]

**Review by rldhont — CHANGES_REQUESTED** [review]

File: lizmap/modules/lizmap/lib/Form/QgisFormControl.php [hunk 2]
```diff
@@ -250,6 +254,11 @@ public function __construct($ref, $properties, $prop, $defaultValue, $constraint
                     if ($max !== null) {
                         $this->ctrl->datatype->addFacet('maxValue', $max);
                     }
+                    // step cast as integer, use only if datatype is integer
+                    $step = $this->getEditAttribute('Step');
+                    if ($step !== null && $this->fieldDataType == 'integer') {
+                        $this->ctrl->setAttribute('stepValue', $step);
+                    }
```
[the remark this answers is not in the document]
Comment (rldhont) [reply lizmap/modules/lizmap/lib/Form/QgisFormControl.php:261]: Documentation about decimal values in input type number https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/number#allowing_decimal_values

[the remark this answers is not in the document]
Comment (nworr) [reply lizmap/modules/lizmap/lib/Form/QgisFormControl.php:261]: stepvalue 0.1 will forbid user to use lower granularity (0.01) . 
special value `any` can be used

[the remark this answers is not in the document]
Comment (rldhont) [reply lizmap/modules/lizmap/lib/Form/QgisFormControl.php:261]: Can we know the field precision ?

[the remark this answers is not in the document]
Comment (nworr) [reply lizmap/modules/lizmap/lib/Form/QgisFormControl.php:261]: only in "Range"  type, so i can set a step value according to it (precision = 4 > step = 0.0001) and `any` elsewhere ?

Any reason why Step is handled as int ? https://github.com/3liz/lizmap-web-client/blob/041cdd72bc190be582065ea7cd2501c447d7b580/lizmap/modules/lizmap/lib/Project/QgisProject.php#L2087

**Comment by rldhont** [comment]
@nworr if the work is ended you can update your branch with master and the PR.

**Comment by nworr** [comment]
@nboisteault can tou take a look to playwright tests ?

**Comment by nboisteault** [comment]
We could also use `pattern="[0-9]"` to even go further.

**Comment by nworr** [comment]
i change end2end tests : 

- playwright test check if `input` element has proper attributes (type=numbrer,  min, max, step)
- cypress test perform form validation tests as if the browser don't support input type number

**Review by rldhont — APPROVED** [review]

**Comment by rldhont** [comment]
Thanks @nworr you can merge at the end of the tests.

**Comment by nworr** [comment]
to be more precise :  the 'step' configuration will be verified by browser only, at this time jelix forms doesn't implement step verification

**Pull request merged**